### PR TITLE
Fix the Magnetic Field in the KFParticle

### PR DESCRIPTION
**Pull request by MYOMAO**
[comment]: <> (Please tell us something about this pull request)

After the studies on KS0 reconstruction performance, we found that the magnetic field used in KFParticle for track propagation is incorrect. The magnetic field is in kilo Gauss instead of Telsa. Also positive magnetic field indicate the +z direction, which is actually where the magnetic field of sPHENIX. Hence, we change the magnetic field from -1.4 -> 14 to represent the B field of 1.4 T in the +z direction for sPHENIX.  

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [ ] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )

## TODOs (if applicable)

[comment]: <> ( In case this is a draft PR, e.g. for running checks using Jenkins, please make the pull request as a draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/  )

### CODE REVIEW

File: offline/packages/KFParticle_sPHENIX/KFParticle_eventReconstruction.cc [hunk 1]
```diff
@@ -56,8 +56,8 @@ void KFParticle_eventReconstruction::createDecay(PHCompositeNode* topNode, std::
                                                  std::vector<std::vector<KFParticle>>& selectedIntermediates,
                                                  int& nPVs, int& multiplicity)
 {
-  //ALICE field is 0.5T but they set it to -5 in KFParticle? Checked momentum sums and -1.4 is more accurate
-  KFParticle::SetField(-1.4e0);
+  //The actual unit of KFParticle is in kilo Gauss (kG), which is equivalent to 0.1 T, instead of Tesla (T). The positive value indicates the B field is in the +z direction
+  KFParticle::SetField(14.0);
```
[the remark this answers is not in the document]
Comment (MYOMAO) [reply offline/packages/KFParticle_sPHENIX/KFParticle_eventReconstruction.cc:60]: I will have a look at that.

[the remark this answers is not in the document]
Comment (osbornjd) [reply offline/packages/KFParticle_sPHENIX/KFParticle_eventReconstruction.cc:60]: [Here](https://github.com/sPHENIX-Collaboration/coresoftware/blob/b68653a3c6eb141d226e86d4e5dc24439c34221f/offline/packages/trackreco/MakeActsGeometry.cc#L478-L483) is where we get the fieldmap. We could get that root file, open it up, and then  get the bz value at (0,0,0).

[the remark this answers is not in the document]
Comment (MYOMAO) [reply offline/packages/KFParticle_sPHENIX/KFParticle_eventReconstruction.cc:60]: It appears that is in T not in kG @osbornjd @blackcathj. We need to convert the unit.

**Review by blackcathj — CHANGES_REQUESTED** [review]
Loading field map in `KFParticle_eventReconstruction::createDecay()` is problematic: 

1. This function is called every event and loading Bfield each time is very inefficient. Please move it to be initialized once. 
2. Please also make the field file name settable via Fun4All macro

A solution is to moving this block of field map reading to `KFParticle_sPHENIX` `InitRun()`, which then set single Bz field value to `KFParticle_eventReconstruction` via a setter like `KFParticle_eventReconstruction::setBz(double B_Tesla)`

**Comment by pinkenburg** [comment]
We should move to using the cdb to get the fieldmap like in 
simulation/g4simulation/g4main/PHG4Reco.cc. This snippet will fall back to CALIBRATIONSROOT if the CDB cannot locate the file:

#include <ffamodules/CDBInterface.h>

char *calibrationsroot = getenv("CALIBRATIONROOT");
  std::string m_magField = "sphenix3dtrackingmapxyz.root";
  if (calibrationsroot != nullptr)
  {
	  m_magField = std::string(calibrationsroot) + std::string("/Field/Map/") + m_magField;
  }
std::string url = CDBInterface::instance()->getUrl("FIELDMAPBIG", m_magField);

**Comment by osbornjd** [comment]
Which fieldmaps are in the CDB? We have several, the big one is way too large for the purposes of tracking so for memory usage purposes we should use the tracking volume only map. Are all of the maps that are in the calibrations repo, i.e. in `$CALIBRATIONROOT/FieldMap/` also in the CDB?

**Comment by pinkenburg** [comment]
Sorry my bad - we need to give it another name "FIELDMAPTRACKING". I'll add it to the cdb and the above construct will read it from CALIBRATIONSROOT until this is done

**Comment by MYOMAO** [comment]
OK I will implement the setter and put my codes there.

**Review by osbornjd — CHANGES_REQUESTED** [review]
Can you also implement Chris' suggestion to call from the CDB if the file exists there, and to otherwise revert to the calibrations repo:

```
#include <ffamodules/CDBInterface.h>
...
char *calibrationsroot = getenv("CALIBRATIONROOT");
std::string m_magField = "sphenix3dtrackingmapxyz.root";

std::string url = CDBInterface::instance()->getUrl("FIELDMAPBIG", m_magField);

if (calibrationsroot != nullptr)
{
m_magField = std::string(calibrationsroot) + std::string("/Field/Map/") + m_magField;
}
```

**Comment by pinkenburg** [comment]
I looked at the cppcheck failure and saw the code does not follow our conventions about include files which are really important
https://wiki.sphenix.bnl.gov/index.php/Codingconventions

**Comment by pinkenburg** [comment]
> Can you also implement Chris' suggestion to call from the CDB if the file exists there, and to otherwise revert to the calibrations repo:
> 
> ```
> #include <ffamodules/CDBInterface.h>
> ...
> char *calibrationsroot = getenv("CALIBRATIONROOT");
> std::string m_magField = "sphenix3dtrackingmapxyz.root";
> 
> std::string url = CDBInterface::instance()->getUrl("FIELDMAPBIG", m_magField);
> 
> if (calibrationsroot != nullptr)
> {
> m_magField = std::string(calibrationsroot) + std::string("/Field/Map/") + m_magField;
> }
> ```

In the meantime I added the tracking fieldmap to our cdb, please use 
std::string url = CDBInterface::instance()->getUrl("FIELDMAPTRACKING", m_magField);
to fetch it. Otherwise you load our big map

**Comment by MYOMAO** [comment]
Somehow I messed up the rebase and produced many commits here...

**Pull request closed without merging**